### DISCUSSION TITLE
How to force IP change?

### DISCUSSION THREAD

**Discussion by ChrizZz90** [qdm12/gluetun#597] [Q&A]
Am I requesting information on a daily basis but the access is limited and can't be requested with the same IP. No problem, I thought, with gluetun I have a swiss knife and find the right tool. My idea is:
1. Setting up a python script to force an IP change, grab the data and store it
2. Using the gluetun HTTP proxy to guide the request through the VPN connection
3. Using the gluetun HTTP Control to stop the VPN connection and start it again

This is the relevant part of my python script:

```py
#Change VPN IP
is_ip = requests.get("http://172.17.0.3:8000/v1/publicip/ip")
print(is_ip.text)
new_ip = is_ip
while new_ip.text == is_ip.text:
  reconnect_stop = requests.put("http://172.17.0.3:8000/v1/openvpn/status", json={"status":"stopped"})
  sleep(10)
  reconnect_start = requests.put("http://172.17.0.3:8000/v1/openvpn/status", json={"status":"running"})
  sleep(15)
  new_ip = requests.get("http://172.17.0.3:8000/v1/publicip/ip")
  print(new_ip.text)
```

It already worked just a few minutes after gluetun started, but It seems like having issues getting a new IP if gluetun is running for a while (6+ hours). 

This is the log if it works:
```
2021/09/02 15:25:23 INFO http server: 200 GET /ip wrote 31B to 172.17.0.3:49628 in 62.555µs
2021/09/02 15:25:23 INFO vpn: stopping
2021/09/02 15:25:23 INFO http server: 200 PUT /status wrote 22B to 172.17.0.3:49628 in 88.577019ms
2021/09/02 15:25:33 INFO vpn: starting
2021/09/02 15:25:33 INFO firewall: setting VPN connection through firewall...
2021/09/02 15:25:33 INFO vpn: OpenVPN 2.5.2 armv7-alpine-linux-musleabihf [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on May  4 2021
2021/09/02 15:25:33 INFO vpn: library versions: OpenSSL 1.1.1l  24 Aug 2021, LZO 2.10
2021/09/02 15:25:33 WARN vpn: --ping should normally be used with --ping-restart or --ping-exit
2021/09/02 15:25:33 INFO vpn: TCP/UDP: Preserving recently used remote address: [AF_INET]IP_hided
2021/09/02 15:25:33 INFO vpn: UDP link local: (not bound)
2021/09/02 15:25:33 INFO vpn: UDP link remote: [AF_INET]IP_hided
2021/09/02 15:25:33 INFO http server: 200 PUT /status wrote 22B to 172.17.0.3:49628 in 306.063466ms
2021/09/02 15:25:33 INFO vpn: [server.prod.surfshark.com] Peer Connection Initiated with [AF_INET]IP_hided
2021/09/02 15:25:34 ERROR vpn: Unrecognized option or missing or extra parameter(s) in [PUSH-OPTIONS]:7: block-outside-dns (2.5.2)
2021/09/02 15:25:34 INFO vpn: TUN/TAP device tun0 opened
2021/09/02 15:25:34 INFO vpn: /sbin/ip link set dev tun0 up mtu 1500
2021/09/02 15:25:34 INFO vpn: /sbin/ip link set dev tun0 up
2021/09/02 15:25:34 INFO vpn: /sbin/ip addr add dev tun0 10.8.8.18/24
2021/09/02 15:25:34 INFO vpn: UID set to nonrootuser
2021/09/02 15:25:34 INFO vpn: Initialization Sequence Completed
2021/09/02 15:25:35 INFO ip getter: Public IP address is IP_new (Country, City)
2021/09/02 15:25:48 INFO http server: 200 GET /ip wrote 30B to 172.17.0.3:49628 in 50.222µs
```

But this is the log if it doesn't work and the python script is stuck in the reconnect loop:

```
2021/09/03 05:30:04 INFO http server: 200 GET /ip wrote 30B to 172.17.0.1:44060 in 158.424µs
2021/09/03 05:30:04 INFO vpn: stopping
2021/09/03 05:30:04 INFO http server: 200 PUT /status wrote 22B to 172.17.0.1:44062 in 107.35656ms
2021/09/03 05:30:14 INFO vpn: starting
2021/09/03 05:30:14 INFO firewall: setting VPN connection through firewall...
2021/09/03 05:30:14 INFO vpn: OpenVPN 2.5.2 armv7-alpine-linux-musleabihf [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on May  4 2021
2021/09/03 05:30:14 INFO vpn: library versions: OpenSSL 1.1.1l  24 Aug 2021, LZO 2.10
2021/09/03 05:30:14 WARN vpn: --ping should normally be used with --ping-restart or --ping-exit
2021/09/03 05:30:14 INFO vpn: TCP/UDP: Preserving recently used remote address: [AF_INET]IP_hided
2021/09/03 05:30:14 INFO vpn: UDP link local: (not bound)
2021/09/03 05:30:14 INFO vpn: UDP link remote: [AF_INET]IP_hided
2021/09/03 05:30:14 INFO http server: 200 PUT /status wrote 22B to 172.17.0.1:44068 in 247.63885ms
2021/09/03 05:30:14 INFO vpn: [server2.prod.surfshark.com] Peer Connection Initiated with [AF_INET]IP_hided
2021/09/03 05:30:15 ERROR vpn: Unrecognized option or missing or extra parameter(s) in [PUSH-OPTIONS]:7: block-outside-dns (2.5.2)
2021/09/03 05:30:15 INFO vpn: TUN/TAP device tun0 opened
2021/09/03 05:30:15 INFO vpn: /sbin/ip link set dev tun0 up mtu 1500
2021/09/03 05:30:15 INFO vpn: /sbin/ip link set dev tun0 up
2021/09/03 05:30:15 INFO vpn: /sbin/ip addr add dev tun0 10.8.8.12/24
2021/09/03 05:30:15 INFO vpn: UID set to nonrootuser
2021/09/03 05:30:15 INFO vpn: Initialization Sequence Completed
###Starting of 2nd loop###
2021/09/03 05:30:29 INFO http server: 200 GET /ip wrote 30B to 172.17.0.1:44078 in 51.462µs
2021/09/03 05:30:29 INFO vpn: stopping
2021/09/03 05:30:29 INFO http server: 200 PUT /status wrote 22B to 172.17.0.1:44080 in 72.624239ms
2021/09/03 05:30:40 INFO vpn: starting
2021/09/03 05:30:40 INFO firewall: setting VPN connection through firewall...
2021/09/03 05:30:40 INFO vpn: OpenVPN 2.5.2 armv7-alpine-linux-musleabihf [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on May  4 2021
2021/09/03 05:30:40 INFO vpn: library versions: OpenSSL 1.1.1l  24 Aug 2021, LZO 2.10
2021/09/03 05:30:40 WARN vpn: --ping should normally be used with --ping-restart or --ping-exit
2021/09/03 05:30:40 INFO vpn: TCP/UDP: Preserving recently used remote address: [AF_INET]IP_hided
2021/09/03 05:30:40 INFO vpn: UDP link local: (not bound)
2021/09/03 05:30:40 INFO vpn: UDP link remote: [AF_INET]IP_hided
2021/09/03 05:30:40 INFO vpn: [server3.prod.surfshark.com] Peer Connection Initiated with [AF_INET]IP_hided
2021/09/03 05:30:40 INFO http server: 200 PUT /status wrote 22B to 172.17.0.1:44090 in 126.405713ms
2021/09/03 05:30:41 ERROR vpn: Unrecognized option or missing or extra parameter(s) in [PUSH-OPTIONS]:7: block-outside-dns (2.5.2)
2021/09/03 05:30:41 INFO vpn: TUN/TAP device tun0 opened
2021/09/03 05:30:41 INFO vpn: /sbin/ip link set dev tun0 up mtu 1500
2021/09/03 05:30:41 INFO vpn: /sbin/ip link set dev tun0 up
2021/09/03 05:30:41 INFO vpn: /sbin/ip addr add dev tun0 10.8.8.12/24
2021/09/03 05:30:41 INFO vpn: UID set to nonrootuser
2021/09/03 05:30:41 INFO vpn: Initialization Sequence Completed
```

And after some loops I guess it was too much for gluetun:

```
2021/09/03 05:34:03 ERROR vpn: RTNETLINK answers: File exists
2021/09/03 05:34:03 INFO vpn: ERROR: Linux route add command failed: external program exited with error status: 2
```

Is it possible that I already have a new IP, but my script is too fast and before "ip getter" can check the external IP am stopping the VPN again? Gluetun is connecting to different surfshark servers, would be surprised if they give me the same IP again.

**Comment by qdm12** [comment]
It might be that completing an openvpn connection usually takes about 6 seconds, but sometimes more. Maybe try setting the 15 seconds to 30 before getting the IP address?

It might also be that some vpn server ip addresses are blacklisted from ip echo services like ipinfo.io so the ip won't be updated.

I'll re-check my code over the weekend on how/when the public ip address is obtained and stored.

The RNETLINK error you see looks like openvpn didn't clean up its ip routes properly on shutdown, I'll have a look. Although personally my vpn (mullvad) blocks me after authenticating too many times so it's rather hard to test 😄 I may get you a Docker image with some debug logs sometimes. Feel free to create a github issue too for that one.

**Comment by ChrizZz90** [comment]
so even with 30sec I am getting the same issue. Even if there is a lot of time between stop and start:

```
2021/09/05 05:30:54 INFO vpn: Initialization Sequence Completed
2021/09/05 05:31:23 INFO http server: 200 GET /ip wrote 31B to 172.20.0.1:37988 in 24.555µs
```

Do you think it makes a difference if I start the script manually with my user or with cron as root?

//edit: That was stupid: Cron is executed as user not root.

**Reply by qdm12** [reply]
Oh ok so does it work as expected in the end?

**Reply by qdm12** [reply]
Also I think that [line](https://github.com/qdm12/gluetun/blob/master/internal/vpn/tunnelup.go#L24) explains it only runs once on the first connection so that's likely why the script doesn't work.

I'll change it to run on every connection.

**Reply by ChrizZz90** [reply]
Yes, this is exactly what I figgured out. I changed the script and checked the IP with an ipify request instead of the gluetun API and it changed but based on the line you mentioned the gluetun API doesn't know that it changed and sends the old IP back.

Now, if I trust the ipify response I it works with the first try. @qdm12 do you need an issue ticket for this?

**Reply by qdm12** [reply]
Actually never mind this comment I put is wrong, the public IP address is fetched on every connection as it should be. Also I'm not sure which version of the program you're using, maybe older ones don't do that. I tried with both Wireguard and OpenVPN and the IP gets updated in the API.

Although sometimes, I had:

```
2021/09/06 13:18:38 INFO vpn: Initialization Sequence Completed
2021/09/06 13:18:53 ERROR ip getter: Get "https://ipinfo.io/ip": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2021/09/06 13:18:53 INFO ip getter: retrying in 5s
2021/09/06 13:19:13 ERROR ip getter: Get "https://ipinfo.io/ip": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2021/09/06 13:19:13 INFO ip getter: retrying in 5s
2021/09/06 13:19:34 WARN ip getter: Get "https://ipinfo.io/89.36.78.136": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2021/09/06 13:19:34 INFO ip getter: Public IP address is 89.36.78.136
```

So it seems ipinfo.io is blocking that particular VPN IP and the retry mechanism takes almost one minute to get the public IP address.
Maybe you get such logs?

What I'll do is clear the IP address once the VPN goes down.

**Reply by qdm12** [reply]
Done in 6b6caa435f56de6bda736ea21856c813f7682e63 in `:latest` image.

Maybe you can modify your script to take into account the special case where the `public_ip` string is empty, meaning it's being obtained, and to wait for it to be filled before sending another PUT request to restart the vpn.

**Reply by ChrizZz90** [reply]
I checked the latest image today but even after 10 minutes, the public_ip string is empty. Is that the expected behaviour?

**Reply by qdm12** [reply]
Are you sure? I have an IP on my side. What commit are you running (at the top of the logs)?

**Reply by ChrizZz90** [reply]
I wasn't :) But now I checked it again with commit 65ace12 and can reproduce it again. I restarted gluetun yesterday and tested it a few minutes ago. Even after 40min the public IP string is empty. It works if I check it directly after a restart.

**Reply by qdm12** [reply]
- 8ac2a816c3399adbffa802ab47de30836aa9213d fixes the Go HTTP client failing to re-use its no longer usable connections after a VPN restart. It now closes all idle connections and that seems to make it work.
- d39201f9b38cfd6bc143a708ca1f096c5ecff52d fixes a deadlock in the public ip loop that was occurring because of the ticker (every 12h by default)

Together they should put a nail in this I hope 😉 
Thanks for taking the time to exchange messages!

**Reply by ChrizZz90** [reply]
brilliant, can confirm, it works now! Thanks!!!

**Reply by qdm12** [reply]
Nice you're welcome!
